### PR TITLE
Add use-resize-observer types

### PR DESCRIPTION
**Pull request by PetrSvirak**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.) 
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/use-resize-observer/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
+// Type definitions for use-resize-observer 5.0
+// Project: https://github.com/ZeeCoder/use-resize-observer#readme
+// Definitions by: PetrSvirak <https://github.com/PetrSvirak>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.5
```
[the remark this answers is not in the document]
Comment (PetrSvirak) [reply types/use-resize-observer/index.d.ts]: I'm not completely sure whether it is [smarter type checking](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-5.html#smarter-union-type-checking) or [better type inference](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-5.html#higher-order-type-inference-from-generic-constructors) or even something else, however, tests for lower TS versions fail with several errors like the following one:
```
ERROR: 12:1  expect  TypeScript@3.4 expected type to be:
  ObserverResultWithoutDefaultSize<HTMLElement>
got:
  ObserverResultWithoutDefaultSize<{}>
```
Fixing this for TS 3.4 and lower breaks tests for TS@next, so I stuck with the 3.5 and newer.

File: types/use-resize-observer/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+export {};
+
+/***
+ * Allows using a ResizeObserver to measure size of an element assigned with ref returned from the hook.
+ * The hook is invoked at least once before the first measurement of actual size by the observer.
+ * @return a ref,
+ *         and ref.current's width (1 before the first measurement),
+ *         and ref.current's height (1 before the first measurement).
+ */
+export default function(): ObserverResultWithDefaultSize;
```
Comment (andrewbranch) [inline types/use-resize-observer/index.d.ts]: The source library has a `main` entry pointing to a CommonJS module and a `module` entry pointing to an ESM module. You’ve correctly modeled the ESM module here. But, TypeScript doesn’t yet have the ability to model both formats side-by-side for the same module, so unfortunately we have to pick one and ignore the other. For now, I think that should be the `main` entry, so this should be written as a CJS module:

```ts
declare function useResizeObserver(): ObserverResultWithDefaultSize;
declare function useResizeObserver(// ... continue with overloads...

export = useResizeObserver;
```

Users can still do `import useResizeObserver from 'use-resize-observer'` if they have `esModuleInterop` turned on.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: types/use-resize-observer/index.d.ts (hunk 2 above)
Comment (PetrSvirak) [reply types/use-resize-observer/index.d.ts]: Makes sense, thanks 😃. 
I did not find a way to export additional interfaces, though 😢. So, I had to remove `ObserverDefaults` and omit exports of `ObserverResult*` from `index.d.ts`. I preserved tests for `ObserverDefaults` even though I had to repeat some types from `index.d.ts` in `the use-resize-observer-tests.ts` – while it is a bit more error-prone, I found the tests quite useful while defining types/expressing possible defaults' variations. See https://github.com/DefinitelyTyped/DefinitelyTyped/commit/e4ef59a31c4cbb0c4ec7f9b3f367cdb77598d10b.

Comment (andrewbranch) [reply types/use-resize-observer/index.d.ts]: Oh, here’s how you do that:

```ts
declare function useResizeObserver(): useResizeObserver.ObserverResultWithDefaultSize;
declare function useResizeObserver(// ... continue with overloads...

declare namespace useResizeObserver {
  export interface ObserverResultWithDefaultSize { ... }
}

export = useResizeObserver;
```

Comment (PetrSvirak) [reply types/use-resize-observer/index.d.ts]: Thanks 👍, interfaces put back in https://github.com/DefinitelyTyped/DefinitelyTyped/commit/370bfedb863aca89cacee499dfef7685cf5c775a.

**Review by andrewbranch — APPROVED** [review]

**Pull request merged**